23-ima-sig-signature-verification.md

```diff
@@ -0,0 +1,287 @@
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed (optional)](#infrastructure-needed-optional)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+<!--
+**ACTION REQUIRED:** In order to merge code into a release, there must be an
+issue in [keylime/enhancements] referencing this enhancement and targeting a release**.
+
+For enhancements that make changes to code or processes/procedures in core
+Keylime i.e., [keylime/keylime], we require the following Release
+Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These
+checklist items _must_ be updated for the enhancement to be released.
+-->
+
+- [ ] Enhancement issue in release milestone, which links to pull request in [keylime/enhancements]
+- [ ] Core members have approved the issue with the label `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place
+- [ ] User-facing documentation has been created in [keylime/keylime-docs]
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+## Summary
+
+<!--
+This section is incredibly important for producing high quality user-focused
+documentation such as release notes or a development roadmap.  It should be
+possible to collect this information before implementation begins in order to
+avoid requiring implementers to split their attention between writing release
+notes and implementing the feature itself. Reviewers
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+-->
+
+Keylime currently only supports verification of IMA measurement lists against
+an allowlist (whitelist) and exclude list. This proposal suggests to extend
+Keylime to support IMA signature verification by supporting the 'ima-sig'
+template.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals and non-goals of
+this enhancement.  Describe why the change is important and the benefits to users.
+-->
+
+The allowlist that is used to verify the IMA measurements received from a
+system may be a huge list with multiple thousand entries. The size of
+the allowlist primarily depends on the number of software packages installed
+on a system and thus the expected number of possible measurements coming from
+that system. An alternative to allowlist reconciliation is to verify file
+signatures using a few keys that were used to sign immutable files on such a
+system. A combination of both is also possible.
+
+Having an exclude list to skip over measurements on mutable files will still be
+necessary since those cannot be signed.
+
```

Comment:
I just want to make sure I understand this better. I read through the IMA docs but didn't quite see the clarification I was looking for. Are the signatures based on the file contents or the hash of the file? Because in order to verify the signature the verifier needs the thing that's being signed right? It only has access to the things in the measurement log which would be the file path, hash and the signature.

Reply:
The signatures are done by signing the hash (sha1, sha256 etc.) over the file. IMA needs to take sha256 measurements for Keylime to be able to validate signatures over a sha256 hash. A sha1 IMA measurement won't allow validation of the signature then, similarly a sha256 measurement but a signature over a sha1 hash won't work.